pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java

```diff
@@ -436,20 +436,22 @@ private void onSendComplete(Throwable e, SendCallback sendCallback, MessageImpl<
             if (payload == null) {
                 log.error("[{}] [{}] Payload is null when calling onSendComplete, which is not expected.",
                         topic, producerName);
-            } else {
-                ReferenceCountUtil.safeRelease(payload);
             }
-            if (e != null) {
-                latencyHistogram.recordFailure(latencyNanos);
-                stats.incrementSendFailed();
-                onSendAcknowledgement(msg, null, e);
-                sendCallback.getFuture().completeExceptionally(e);
-            } else {
-                latencyHistogram.recordSuccess(latencyNanos);
-                publishedBytesCounter.add(msgSize);
-                stats.incrementNumAcksReceived(latencyNanos);
-                onSendAcknowledgement(msg, msg.getMessageId(), null);
-                sendCallback.getFuture().complete(msg.getMessageId());
+            try {
+                if (e != null) {
+                    latencyHistogram.recordFailure(latencyNanos);
+                    stats.incrementSendFailed();
+                    onSendAcknowledgement(msg, null, e);
+                    sendCallback.getFuture().completeExceptionally(e);
+                } else {
+                    latencyHistogram.recordSuccess(latencyNanos);
+                    publishedBytesCounter.add(msgSize);
+                    stats.incrementNumAcksReceived(latencyNanos);
+                    onSendAcknowledgement(msg, msg.getMessageId(), null);
+                    sendCallback.getFuture().complete(msg.getMessageId());
+                }
+            } finally {
+                ReferenceCountUtil.safeRelease(payload);
```

Comment:
Could we remove the null check for payload?

Reply:
I assume you are talking the code below:
```java
            if (payload == null) {
                log.error("[{}] [{}] Payload is null when calling onSendComplete, which is not expected.",
                        topic, producerName);
            }
```

This seems harmless, I have no clue about original intention here, so let's keep this patch simple and stupid :->

Reply:
This patch change the original logic.

```
if (payload == null) {
                log.error("[{}] [{}] Payload is null when calling onSendComplete, which is not expected.",
                        topic, producerName);
            } else {
                ReferenceCountUtil.safeRelease(payload);
            }
```
Before this patch, only payload is not null we will release it, while after this patch, the null check is removed.

Reply:
Oh, `ReferenceCountUtil.safeRelease` is null-safe as its name suggests. It does nothing if invoked with a `null` value, and this behavior is explicitly documented in its Javadoc. 

https://github.com/netty/netty/blob/c7de17a7f385a8a2156ddc9e74c7a85a6651a034/common/src/main/java/io/netty/util/ReferenceCountUtil.java#L114